### PR TITLE
PD-129: initial parsing for json filters

### PR DESCRIPTION
**Pull request by kamaurra-headlight**
* [PD-129](https://teamheadlight.atlassian.net/browse/PD-129)

### CODE REVIEW

File: source/Meadow-Filter-Parse.js [hunk 1]
```diff
@@ -140,6 +265,63 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
...
+			break;
+		}
+
+		case 'FSJF':   // Filter Sort JSON Field
+		{
+			const { fieldColumn, jsonPath } = parseJSONFieldAndPath(pFilterStanza);
+			const sortDirection = (pFilterStanza.Operator === 'DESC') ? 'Descending' : 'Ascending';
+			const sortingType = getDataType(pFilterStanza.Value);
+			// NOTE: bypassing the table name injection by not providing a column argument
+			// TODO: prefix 'fieldColumn' with table name to avoid ambiguity (e.g. in joins)
```
Comment (kamaurra-headlight) [inline source/Meadow-Filter-Parse.js:333]: for existing filters, we don't handle this either (e.g. see FBD).

File: source/Meadow-Filter-Parse.js [hunk 2]
```diff
@@ -100,6 +131,100 @@ const getFilterComparisonOperator = (pFilterOperator) =>
...
+ * @param {string} connector - the logical connector to this query (AND/OR)
+ * @param {object} pQuery - the foxhound query object
+ * @return {object} pQuery
+ */
+const addFilterJSONToQuery = (fieldColumn, jsonPath, value, comparisonOperator, connector, pQuery) =>
+{
+	pQuery.addFilter('', '', '(');
+	// NOTE: bypassing the table name injection by not providing a column argument
+	// TODO: prefix 'fieldColumn' with table name to avoid ambiguity (e.g. in joins)
+	pQuery.addFilter('', 1, `JSON_VALID(${fieldColumn}) =`, 'AND', 'validJson');
```
Comment (kamaurra-headlight) [inline source/Meadow-Filter-Parse.js]: for this workaround, I've cut a ticket for adding more explicit support for this in foxhound:
https://teamheadlight.atlassian.net/browse/PD-328

File: source/Meadow-Filter-Parse.js [hunk 3]
```diff
@@ -45,6 +45,37 @@
...
+ FBJL - Filter By JSON List (value list, separated by commas)
+       FBJL~FormDataJSON.Tidings.Data.Type~INN~ChangeOrder,Payment Form
+ FBJLOR - Filter By JSON List (value list, separated by commas, left-side OR connected)
+       FBJLOR~FormDataJSON.Tidings.Data.Type~INN~ChangeOrder,Payment Form
+ FBJD - Filter by JSON Date (exclude time)
+       FBJD~FormDataJSON.UpdateDate~EQ~2015-10-01
+ FSJF - Filter Sort JSON Field
+       FSJF~FormDataJSON.Category~ASC~0
+       FSJF~FormDataJSON.Category~DESC~<cast>
+          where <cast> can be in (UINT,SINT,CHAR,DD,DATE) - defaults to CHAR
```
Comment (bigal-hl) [inline source/Meadow-Filter-Parse.js:78]: We should have a legend somewhere in this documentation for these types.

If someone uses this, what ends up happening if a value in a record doesn't match the requested type?

File: source/Meadow-Filter-Parse.js [hunk 4]
```diff
@@ -140,6 +265,63 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
...
+			break;
+		}
+
+		case 'FBJD':   // Filter by JSON Date (exclude time)
+		{
+			const { fieldColumn, jsonPath } = parseJSONFieldAndPath(pFilterStanza);
+			pQuery.addFilter('', '', '(');
+			// NOTE: bypassing the table name injection by not providing a column argument
+			// TODO: prefix 'fieldColumn' with table name to avoid ambiguity (e.g. in joins)
+			pQuery.addFilter('', 1, `JSON_VALID(${fieldColumn}) =`, 'AND', 'validJson');
```
Comment (bigal-hl) [inline source/Meadow-Filter-Parse.js]: Is this broken in foxhound, or is this caused by the API code here? https://github.com/paviasystems/mpi_api/blob/master/ts_source/controller/HeadlightController.ts#L555-L566

If the latter, it seems like we could just fix this there, and not need this hack.

File: source/Meadow-Filter-Parse.js [hunk 5]
```diff
@@ -100,6 +131,100 @@ const getFilterComparisonOperator = (pFilterOperator) =>
...
+ * @param {object} pQuery - the foxhound query object
+ * @return {object} pQuery
+ */
+const addFilterJSONToQuery = (fieldColumn, jsonPath, value, comparisonOperator, connector, pQuery) =>
+{
+	pQuery.addFilter('', '', '(');
+	// NOTE: bypassing the table name injection by not providing a column argument
+	// TODO: prefix 'fieldColumn' with table name to avoid ambiguity (e.g. in joins)
+	pQuery.addFilter('', 1, `JSON_VALID(${fieldColumn}) =`, 'AND', 'validJson');
+	let command = `JSON_EXTRACT(${fieldColumn}, '$${jsonPath}')`;
```
Comment (bigal-hl) [inline source/Meadow-Filter-Parse.js:216]: Is there a way to abuse this by include quotes in the JSON path?

File: source/Meadow-Filter-Parse.js [hunk 6]
```diff
@@ -100,6 +131,100 @@ const getFilterComparisonOperator = (pFilterOperator) =>
...
+			finalValue = value.map((v) => Number(v));
+		}
+	}
+	else if (!isNaN(Number(value)))
+	{
+		// assumption: request contained a number embedded in a string, number comparison is desired
+		// value is a number, extract it from the strings
+		command = `JSON_UNQUOTE(${command})`;
+		finalValue = Number(value);
+	}
```
Comment (bigal-hl) [inline source/Meadow-Filter-Parse.js]: This is a little hard to follow, and definitely needs a clearer comment, IMO. What is this for, exactly?

File: source/Meadow-Filter-Parse.js [hunk 7]
```diff
@@ -100,6 +131,100 @@ const getFilterComparisonOperator = (pFilterOperator) =>
...
+	// NOTE: for whatever reason, foxhound accepts the JSON_UNQUOTE/JSON_EXTRACT commands without the bypass needed for JSON_VALID
```
Comment (bigal-hl) [inline source/Meadow-Filter-Parse.js]: This seems sus. Would like to understand this better. What I observed when working on this with Anu was that the workaround was not needed if the field was fully qualified (had a `.` in it), which is why I am suspicious about that "qualifying" API code.

File: source/Meadow-Filter-Parse.js (hunk 3 above)
Comment (kamaurra-headlight) [reply source/Meadow-Filter-Parse.js:78]: MySQL returns a default value for that type. Example with JCCard JCSequenceNumber, where some sequence numbers are actually alphanumeric:
`JCSeqNumCastToDecimal`
0.00000

`Original/Raw JCSequence Number`
JC 621003`

File: source/Meadow-Filter-Parse.js (hunk 6 above)
Comment (kamaurra-headlight) [reply source/Meadow-Filter-Parse.js]: If the filter value is a number, then we make sure to unquote that filter value and the target field to ensure proper comparison.
This solves cases like below, where you want the JCCards with a Sequence Number greater than 10:
```
FBJV~FormDataJSON.Header.JCSequenceNumber~GT~10
```

without this casting and unquoting, SQL would compare values in the DB with string "10" instead of number 10.

File: source/Meadow-Filter-Parse.js (hunk 4 above)
Comment (kamaurra-headlight) [reply source/Meadow-Filter-Parse.js]: oh! I think it is. Nice, will look into cleaning that up

File: source/Meadow-Filter-Parse.js (hunk 5 above)
Comment (kamaurra-headlight) [reply source/Meadow-Filter-Parse.js:216]: YES, there is! Should foxhound ultimately guard against any injections?

Comment (kamaurra-headlight) [reply source/Meadow-Filter-Parse.js:216]: I'm able to get the backend to execute extra SQL queries, specifically SELECT queries. The WHERE clause just errors out (SQL syntax error) if I try to inject an UPDATE or DELETE operation.

File: source/Meadow-Filter-Parse.js (hunk 7 above)
Comment (kamaurra-headlight) [reply source/Meadow-Filter-Parse.js]: I do think it's part of that API code you linked. I'm going to patch that today and see.

File: source/Meadow-Filter-Parse.js [hunk 8]
```diff
@@ -100,6 +131,115 @@ const getFilterComparisonOperator = (pFilterOperator) =>
...
+ * @param {object} pFilterStanza - a filter stanza with the properties defined in the interface
+ * @returns {object} structure { fieldColumn, jsonPath }
+ */
+const parseJSONFieldAndPath = (pFilterStanza) =>
+{
+	const splitIndex = pFilterStanza.Field.indexOf('.');
+	if (splitIndex < 1)
+	{
+		const msg = `Invalid format for Field[${pFilterStanza.Field}]`;
+		throw new Error(msg);
```
Comment (bigal-hl) [inline source/Meadow-Filter-Parse.js:188]: I think this is the main remaining open question in my mind - does it make sense for this to throw? Will that be handled well in the existing layers?

**Review by bigal-hl — APPROVED** [review]

File: source/Meadow-Filter-Parse.js (hunk 8 above)
Comment (kamaurra-headlight) [reply source/Meadow-Filter-Parse.js:188]: >  does it make sense for this to throw?

Well, it was the simplest way to flag the error and ensure that we did not execute a partial query (that would most likely return more data than it should).
That said, we could also add a filter that always evaluates to false so it returns no records instead. But then it's a bit ambiguous (is it an error or is there actually no data for my request?).

> Will that be handled well in the existing layers?

I believe so. Integration test validates this is handled by API. Our micro-services have the handler catch-all, which should also capture this.

**Comment by kamaurra-headlight** [comment]
Follow-up ticket: https://teamheadlight.atlassian.net/browse/PD-359
